internal/installcfg/builder/builder.go

```diff
@@ -132,52 +133,93 @@ func (i *installConfigBuilder) getBasicInstallConfig(cluster *common.Cluster) (*
 		}
 	}
 
-	if i.mirrorRegistriesBuilder.IsMirrorRegistriesConfigured() {
-		isOpenShiftVersionRecentEnough, err := common.BaseVersionGreaterOrEqual(minimalOpenShiftVersionForImageDigestSupport, cluster.OpenshiftVersion)
-		if err != nil {
-			return nil, err
-		}
-		if isOpenShiftVersionRecentEnough {
-			err := i.setImageDigestMirrorSet(cfg)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			// If version does not support ImageDigestSources, set ImageContent
-			err := i.setImageContentSources(cfg)
-			if err != nil {
-				return nil, err
-			}
-		}
+	if err := i.handleMirrorRegistry(cfg, cluster); err != nil {
+		return nil, err
 	}
 
 	return cfg, nil
 }
 
-func (i *installConfigBuilder) setImageDigestMirrorSet(cfg *installcfg.InstallerConfigBaremetal) error {
-	mirrorRegistriesConfigs, err := i.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
+func (i *installConfigBuilder) handleMirrorRegistry(cfg *installcfg.InstallerConfigBaremetal, cluster *common.Cluster) error {
+	isOpenShiftVersionRecentEnough, err := common.BaseVersionGreaterOrEqual(minimalOpenShiftVersionForImageDigestSupport, cluster.OpenshiftVersion)
 	if err != nil {
-		i.log.WithError(err).Errorf("Failed to get the mirror registries conf need for ImageDigestSources")
 		return err
 	}
-	imageDigestSourceList := make([]installcfg.ImageDigestSource, len(mirrorRegistriesConfigs))
-	for i, mirrorRegistriesConfig := range mirrorRegistriesConfigs {
-		imageDigestSourceList[i] = installcfg.ImageDigestSource{Source: mirrorRegistriesConfig.Location, Mirrors: mirrorRegistriesConfig.Mirror}
+	configuration, err := cluster.GetMirrorRegistryConfiguration()
+	if err != nil {
+		return err
 	}
-	cfg.ImageDigestSources = imageDigestSourceList
+
+	i.log.Infof("found OpenShift version %s, setting mirror registry %+v", cluster.OpenshiftVersion, configuration)
+	if isOpenShiftVersionRecentEnough {
+		if err = i.setImageDigestMirrorSet(cfg, configuration); err != nil {
+			return err
+		}
+	} else {
+		// If version does not support ImageDigestSources, set ImageContent
+		if err = i.setImageContentSources(cfg, configuration); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
-func (i *installConfigBuilder) setImageContentSources(cfg *installcfg.InstallerConfigBaremetal) error {
-	mirrorRegistriesConfigs, err := i.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
-	if err != nil {
-		i.log.WithError(err).Errorf("Failed to get the mirror registries conf need for ImageContentSources")
-		return err
+func (i *installConfigBuilder) setImageDigestMirrorSet(cfg *installcfg.InstallerConfigBaremetal, configuration *v1beta1.MirrorRegistryConfiguration) error {
```

Comment:
Is there a way we can deduplicate these functions?

Reply:
Not really; both methods, `setImageContentSources` and `setImageDigestMirrorSet` loop over `ImageDigestMirrors` and set different elements within this loop. One sets the `imageContentSourceList`, and the other sets the `imageDigestSourceList`.
If we want to combine them, we would need to loop over this data twice—once for extracting the data and another time for setting it.
In my opinion, this approach is better and more readable than the combined one.